src/library_manager/llext_manager_dram.c

```diff
@@ -0,0 +1,311 @@
+
+	lib_manager_dram.ctx = rmalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_L3,
+				       buf_size);
+	if (!lib_manager_dram.ctx)
+		return -ENOMEM;
+
+	lib_manager_dram.mod = (struct lib_manager_module *)(lib_manager_dram.ctx + n_lib);
+	lib_manager_dram.sect = (struct llext_elf_sect_map *)(lib_manager_dram.mod + n_mod);
+	lib_manager_dram.llext = (struct llext *)(lib_manager_dram.sect + n_sect);
+	lib_manager_dram.bldr = (struct llext_buf_loader *)(lib_manager_dram.llext + n_llext);
+	lib_manager_dram.sym = (struct llext_symbol *)(lib_manager_dram.bldr + n_llext);
+
+	tr_dbg(&lib_manager_tr, "backup %u libs of %u modules with %u LLEXT with %u sections",
+	       n_lib, n_mod, n_llext, n_sect);
+
+	tr_dbg(&lib_manager_tr, "backup %p to %p, mod %p, loader %p",
+	       lib_manager_dram.ctx, (void *)((uint8_t *)lib_manager_dram.ctx + buf_size),
+	       lib_manager_dram.mod, lib_manager_dram.bldr);
+
+	/* Walk all libraries */
+	for (i = 0, j = 0, l = 0, n_mod = 0, n_sect = 0, n_sym = 0;
+	     i < ARRAY_SIZE(_ext_lib->desc); i++) {
+		if (!_ext_lib->desc[i])
+			continue;
+
+		struct lib_manager_module *mod = _ext_lib->desc[i]->mod;
+
+		/* Copy all modules in each library */
+		lib_manager_dram.ctx[j++] = *_ext_lib->desc[i];
+		memcpy(lib_manager_dram.mod + n_mod, mod,
+		       sizeof(lib_manager_dram.mod[0]) * _ext_lib->desc[i]->n_mod);
+		tr_dbg(&lib_manager_tr, "lib %u base %p", j - 1,
+		       lib_manager_dram.ctx[j - 1].base_addr);
+		n_mod += _ext_lib->desc[i]->n_mod;
+
+		/* Copy instantiated extensions */
+		for (k = 0; k < _ext_lib->desc[i]->n_mod; k++) {
+			if (!mod[k].llext)
+				continue;
+
+			tr_dbg(&lib_manager_tr, "mod %u of %u sections", k,
+			       mod[k].llext->sect_cnt);
+
+			/* Copy the extension and the loader */
+			lib_manager_dram.llext[l] = *mod[k].llext;
+			lib_manager_dram.bldr[l] = *mod[k].ebl;
+
+			/* Copy the section map */
+			memcpy(lib_manager_dram.sect + n_sect, mod[k].ebl->loader.sect_map,
+			       mod[k].llext->sect_cnt * sizeof(lib_manager_dram.sect[0]));
+			n_sect += mod[k].llext->sect_cnt;
+
+			/* Copy exported symbols */
+			if (mod[k].llext->exp_tab.sym_cnt) {
+				memcpy(lib_manager_dram.sym + n_sym, mod[k].llext->exp_tab.syms,
+				       mod[k].llext->exp_tab.sym_cnt *
+				       sizeof(lib_manager_dram.sym[0]));
+				lib_manager_dram.llext[l].exp_tab.syms = lib_manager_dram.sym +
+					n_sym;
+				n_sym += mod[k].llext->exp_tab.sym_cnt;
+			}
+
+			l++;
+		}
+	}
+
+	/* Also flatten dependency lists */
+	int ret = llext_relink_dependency(lib_manager_dram.llext, n_llext);
+
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "Inconsistent dependencies!");
+		return ret;
+	}
+
+	lib_manager_dram.n_llext = n_llext;
+	dcache_writeback_region((__sparse_force void __sparse_cache *)&lib_manager_dram,
+				sizeof(lib_manager_dram));
+	dcache_writeback_region((__sparse_force void __sparse_cache *)lib_manager_dram.ctx,
+				buf_size);
+
+	return 0;
+}
+
+int llext_manager_restore_from_dram(void)
```

Comment:
save and restore could do with more inline comments.

Reply:
@lgirdwood I never succeed at this :-D I knew there would be perfectly reasonable requests for adequate comments, so I made an extra effort to comment everything that I could think of as deserving comments :-)

Reply:
I general best to inline comment a code block that is over 10 line or makes decisions based on external (to this file/func) assumptions that a reviewer will need to search for to understand. If in doubt add a comment :)